src/SubVerso/Highlighting/Code.lean

```diff
@@ -670,16 +674,36 @@ partial def closeUntil (pos : String.Pos) : HighlightM Unit := do
 
   if more then closeUntil pos
 
+/-- Records the corresponding source position of the syntax most recently added to the output. -/
+def setLastPos (lastPos? : Option String.Pos) : HighlightM Unit := do
+  modify fun st => { st with lastPos? }
+
+/-- Adds to the output any source (if any exists) lying between the last added position and `pos`. -/
+def fillMissingSourceUpTo (pos : String.Pos) : HighlightM Unit := do
+  if let some lastPos := (← get).lastPos? then
+    if lastPos < pos then
+      let text ← getFileMap
+      openUntil <| text.toPosition lastPos
```

Comment:
I think parser errors may need a little more clever handling here. They may occur in a substring of the unparsed region, after all. I think this code should probably segment the unparsed region according to the messages that occur in it, and emit it accordingly. Does that make sense?